pyogrio/tests/test_raw_io.py

```diff
@@ -390,16 +378,27 @@ def test_read_write_data_types_numeric(tmp_path, ext):
 
     # other integer data types that don't roundtrip exactly
     # these are generally promoted to a larger integer type except for uint64
-    for i, (dtype, result_dtype) in enumerate([
-        ("int8", "int16"),
-        ("uint8", "int16"),
-        ("uint16", "int32"),
-        ("uint32", "int64"),
-        ("uint64", "int64")
-    ]):
+    for i, (dtype, result_dtype) in enumerate(
+        [
+            ("int8", "int16"),
+            ("uint8", "int16"),
+            ("uint16", "int32"),
+            ("uint32", "int64"),
+            ("uint64", "int64"),
+        ]
+    ):
         field_data = [np.array([1, 2, 3], dtype=dtype)]
         filename = tmp_path / f"test{i}.{ext}"
         write(filename, geometry, field_data, ["col"], **meta)
         result = read(filename)[3][0]
         assert np.array_equal(result, np.array([1, 2, 3]))
         assert result.dtype == result_dtype
+
+
+def test_read_data_types_numeric_with_null(test_gpkg_nulls):
+    fields = read(test_gpkg_nulls)[3]
+
+    # all fields should be cast to float64 and last value should be np.nan
+    for field in fields:
+        assert field.dtype == "float64"
```

Comment:
float32 should be preserved as float32?

Reply:
I had mistakenly assumed that dtypes would be preserved from GeoPandas through Fiona to the GPKG, but this appears not to be the case (happens even if there are no nulls set when writing to Fiona); all integer fields are converted to `int64` regardless of nulls.  And had also assumed that we needed `float64` to hold `np.nan`, but that isn't true either.

The `float32` field is indeed preserved correctly according to `ogrinfo`, but apparently we're not detecting the float32 subtype correctly in this case (it appears not to be set, according to logging I've added).  Investigating now...

Reply:
Oh - nevermind, `ogrinfo` is showing both as `float64` (`Real`) as well; was misled by our column names with the dtypes again...

I think the upshot is that we need to do something else to make sure dtypes are correctly written.  Updated to write the non-null GPKG ourselves then use Fiona to append in nulls; appears to preserve GDAL equivalents of our field types (none of unsigned integer types supported) and also preserves float32.